lib/services/role.go

```diff
@@ -1185,6 +1288,98 @@ func (set RoleSet) CheckAccessToKubernetes(namespace string, kube *KubernetesClu
 	return trace.AccessDenied("access to kubernetes cluster denied")
 }
 
+// CheckAccessToDatabaseServer checks if this role set has access to the
+// specified database server.
+//
+// Used to filter available databases a user sees with "tsh db ls" command.
+func (set RoleSet) CheckAccessToDatabaseServer(server DatabaseServer) error {
+	var errs []error
+	// Check deny rules.
+	for _, role := range set {
+		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Deny), server.GetNamespace())
+		matchLabels, labelsMessage, err := MatchLabels(role.GetDatabaseLabels(Deny), server.GetAllLabels())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matchNamespace && matchLabels {
+			log.WithField(trace.Component, teleport.ComponentRBAC).Debugf(
+				"Access to database %q denied, deny rule in %q matched; match(namespace=%v, label=%v).",
+				server.GetName(), role.GetName(), namespaceMessage, labelsMessage)
+			return trace.AccessDenied("access to database denied")
+		}
+	}
+	// Check allow rules.
+	for _, role := range set {
+		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Allow), server.GetNamespace())
+		matchLabels, labelsMessage, err := MatchLabels(role.GetDatabaseLabels(Allow), server.GetAllLabels())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matchNamespace && matchLabels {
+			log.WithField(trace.Component, teleport.ComponentRBAC).Debugf(
+				"Access to database %q granted, allow rule in %q matched; match(namespace=%v, label=%v).",
+				server.GetName(), role.GetName(), namespaceMessage, labelsMessage)
+			return nil
+		}
+		if log.GetLevel() == log.DebugLevel {
+			deniedError := trace.AccessDenied("role=%v, match(namespace=%v, label=%v)",
+				role.GetName(), namespaceMessage, labelsMessage)
+			errs = append(errs, deniedError)
+		}
+	}
+	log.WithField(trace.Component, teleport.ComponentRBAC).Debugf(
+		"Access to database %q denied, no allow rule matched; %v.", server.GetName(), errs)
+	return trace.AccessDenied("access to database denied")
+}
+
+// CheckAccessToDatabase checks if this role set has access to a particular
+// database and database user within the specified database proxy.
+//
+// Used as an authorization check when a user connects to a database.
+func (set RoleSet) CheckAccessToDatabase(server DatabaseServer, dbName, dbUser string) error {
+	var errs []error
+	// Check deny rules.
+	for _, role := range set {
+		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Deny), server.GetNamespace())
+		matchLabels, labelsMessage, err := MatchLabels(role.GetDatabaseLabels(Deny), server.GetAllLabels())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		matchName, nameMessage := MatchDatabaseName(role.GetDatabaseNames(Deny), dbName)
+		matchUser, userMessage := MatchDatabaseUser(role.GetDatabaseUsers(Deny), dbUser)
+		if matchNamespace && matchLabels && (matchName || matchUser) {
```

Comment:
I think labels, name and user checks should all happen separately, so you can mix-and-match labels from some roles and names from others.

If I create a role that has:
```
deny:
  dbUsers: "admin"
```
it won't actually deny anything because there are no `dbLabels`

Similarly, If I create one role that grants access to some database labels and a restricted db username, and also create a second role that lists no databases but adds a more privileged db username, I expect the two to merge and grant access using the privileged username.

Reply:
Hmm, are you sure this is how it's supposed to work? I was under the impression that we treat each role as a separate entity and check against each role individually, not merge them. Otherwise I think it may become quite hard to reason about them. @russjones What do you think?

Reply:
The more I think about it the more I'm convinced that we shouldn't merge the roles TBH, as it can potentially lead to dangerous outcomes I think. For example, if I create a role that specifically grants access to a privileged database account in a particular database, I wouldn't expect it to also grant access to a privileged database account in a different role's database, which would happen if we merge them. So IMHO the way it's implemented now follows the principle of "least surprise" and "least privilege".

Reply:
It makes sense. To avoid the other kind of surprise (setting dbUsers/dbNames restrictions on a role, but not labels), how about we make it an error to set dbUsers and dbNames rules on a role unless dbLabels is also set explicitly?

Reply:
This is a good suggestion, I'll do it.

Reply:
Actually, looking at Role's CheckAndSetDefaults function, it defaults to wildcard labels if they're not set (for all nodes/apps/dbs) so unless I'm missing something you shouldn't be able to create a role with empty labels:

https://github.com/gravitational/teleport/blob/roman/dba/api/types/role.go#L530-L532

Reply:
Ah, nice.
It only defaults for `Allow` rules, but not `Deny` rules.
Add defaulting for `Deny` labels, if any other db `Deny` fields are set

Reply:
Done.